pkg/trace/api/api.go

```diff
@@ -210,9 +212,20 @@ func (r *HTTPReceiver) replyTraces(v Version, w http.ResponseWriter) {
 
 // handleTraces knows how to handle a bunch of traces
 func (r *HTTPReceiver) handleTraces(v Version, w http.ResponseWriter, req *http.Request) {
+	if atomic.LoadInt64(&r.refuse) == 1 {
+		// using too much memory
+		io.Copy(ioutil.Discard, req.Body)
+		w.WriteHeader(http.StatusNotAcceptable)
+		io.WriteString(w, fmt.Sprintf("request rejected; trace-agent is past memory threshold (apm_config.max_memory: %.0f bytes)", r.conf.MaxMemory))
+		metrics.Count("datadog.trace_agent.receiver.payload_refused", 1, []string{"reason:mem"}, 1)
+		return
+	}
 	if !r.PreSampler.Sample(req) {
+		// using too much CPU
 		io.Copy(ioutil.Discard, req.Body)
-		r.replyTraces(v, w)
+		w.WriteHeader(http.StatusNotAcceptable)
```

Comment:
Why not `HTTP 429 Too Many Requests`? That could be more descriptive

Reply:
You're right. We should use that for the final version. But for now I think this should be fine